File: 0739-daily-temperatures/0739-daily-temperatures.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;
        int[] ans = new int[n];
        int[] stack = new int[n];
        int top = -1;

        for (int i = 0; i < n; i++) {
            while (top >= 0 && temperatures[i] > temperatures[stack[top]]) {
                int idx = stack[top--];
                ans[idx] = i - idx;
            }
            stack[++top] = i;
        }

        return ans;
    }
}

/*
 0. 1. 2. 3. 4. 5. 6. 7
[73,74,75,71,69,72,76,73]
 */

File: 0739-daily-temperatures/0739-daily-temperatures.cpp
/*
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& t) {
        ios_base::sync_with_stdio(false);
        cin.tie(NULL);
        cout.tie(NULL);

        int n = t.size();
        vector<int> ans(n, 0);

        for (int currDay = n - 2; currDay >= 0; currDay--) {
            int futureDay = currDay + 1;

            while (futureDay < n && t[currDay] >= t[futureDay]) {
                if (ans[futureDay] == 0) {
                    futureDay = n;
                    break;
                }
                futureDay += ans[futureDay];
            }

            if (futureDay < n) {
                ans[currDay] = futureDay - currDay;
            }
        }

        return ans;
    }
};
*/
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& t) {
        ios_base::sync_with_stdio(false);
        cin.tie(NULL);
        cout.tie(NULL);

        int n = t.size();
        vector<int> ans(n, 0);

        int* st = new int[n];
        int top = -1;

        for (int currDay = 0; currDay < n; currDay++) 
        {
            while (top >= 0 && t[currDay] > t[st[top]]) {
                int pastIdx = st[top];
                ans[pastIdx] = currDay - pastIdx;
                top--;
            }

            top++;
            st[top] = currDay;

        }

        return ans;
    }
};